File: leetcode/0671-second-minimum-node-in-a-binary-tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int rootVal;

    public int findSecondMinimumValue(TreeNode root) {
        if (root == null) {
            return -1;
        }
        rootVal = root.val;
        return dfs(root);
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return -1;
        }
        if (node.val != rootVal) {
            return node.val;
        }
        int left = dfs(node.left);
        int right = dfs(node.right);
        if (left == -1) {
            return right;
        }
        if (right == -1) {
            return left;
        }
        return Math.min(left, right);
    }
}

File: leetcode/0671-second-minimum-node-in-a-binary-tree/solution.cpp
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        if (root == nullptr) {
            return -1;
        }
        rootVal = root->val;
        return dfs(root);
    }

private:
    int rootVal;

    int dfs(TreeNode* node) {
        if (node == nullptr) {
            return -1;
        }
        if (node->val != rootVal) {
            return node->val;
        }
        int left = dfs(node->left);
        int right = dfs(node->right);
        if (left == -1) {
            return right;
        }
        if (right == -1) {
            return left;
        }
        return min(left, right);
    }
};
